Wizard notice to notification

JesserH opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* Remove the wizard notice on the dashboard and refactor it into a notification in the notification centre.

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another repo, start you changelog item with the repo name between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/repos, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Creates a new class `class-wizard-notification` which can be accessed to get the appropriate notification depending on the state. Also removes the wizard notice completely

## Relevant technical choices:

**IMPORTANT NOTICE
There is no set date when the admin-redesign branch is merged. Therefore, the version for this upgrade routine in `class-upgrade.php` is set to 14.4 because it will probably be merged around that time, however this may change in the future.**

* The notification is built up with 3 states:
  0 - The wizard has not been started and the notification is a call to action to open it
  1 - The wizard has been opened but has not been finished, the notification is a call to action to 
  open it
  2 - The wizard has been finished and the notification hints that you can reopen it to change 
  some settings with a call to action
The class `class-wizard-notification` takes care of these notifications
* The old function notification in `class-configuration-notifier.php` has been removed which styled the wizard notice before but is not applicable in this approach

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
This PR can be tested by following these steps:

Testing if the old unused db meta is removed:
**QA can skip the first step, please double check if this version is correct (14.4-RC0) **
1. In class-upgrade.php, change line 56 to `'14.4-RC0'   => 'upgrade_144',` (line 56 is the )
2. Checkout this branch, build and activate the plugin
3. Install and configure the yoast test helper
4. Open the yoast test helper and make sure the db version is set to `14.0-RC4`
5. Open a DB browser (i used sequel pro), and in the table `wp_usermeta` search for a field with `meta_key` to have the value of `wpseo-dismiss-configuration-notice`
6. Make sure it exists
7. Go to the yoast test helper again, and set the db version to `13.0-RC0` and hit save
8. Open the db browser again and make sure the field searched for earlier is gone

Testing the functionality:
Step 1 and 2 below are already done in the above test instructions, but are mentioned just in case.
1. Checkout this branch, build and activate the plugin
2. Install and configure the yoast test helper
  **Update by Andrea:** the test helper plugin needs to be switched to the `configuration-wizard-update` branch.
3. Go to Tools-> Yoast Test and click the "Reset Configuration Wizard" button
4. Go to the Yoast SEO plugin dashboard
5. Check if you have the following notification:
![Screenshot 2020-04-28 at 10 45 24](https://user-images.githubusercontent.com/9864796/80466868-60e74100-893d-11ea-96f4-2e58992a3aa7.png)
6. Open the configuration wizard and click the `CLOSE THE WIZARD` button
7. Check if the above notification is gone and you now have this notification:
![Screenshot 2020-04-24 at 12 56 50](https://user-images.githubusercontent.com/9864796/80205538-127b2f00-862b-11ea-92b8-17342becba3a.png)
8. Open the wizard again, go to the last page and click the `Close` button
9. You are redirected to the dashboard, refresh that page
10. Check if the above notifications are gone and you now have the following notification:
![Screenshot 2020-04-24 at 12 58 55](https://user-images.githubusercontent.com/9864796/80205712-5d954200-862b-11ea-9837-f0abbc223f3b.png)


## UI changes
* [X] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation
* [X] I have written documentation for this change.

## Quality assurance

* [X] I have tested this code to the best of my abilities
* [ ] I have added unittests to verify the code works as intended

Fixes #14800

afercia commented:
Note: this PR should have been based on the `feature/admin-redesign` branch.

afercia commented:
Rebased against `feature/admin-redesign`. Also reverted changes to `yarn.lock`.

afercia commented:
The notifications in the new admin-redesigned notification center:

<img width="725" alt="Screenshot 2020-04-29 at 13 35 31" src="https://user-images.githubusercontent.com/1682452/80591947-e0990c80-8a1e-11ea-9775-291afee127eb.png">

<img width="719" alt="Screenshot 2020-04-29 at 13 37 08" src="https://user-images.githubusercontent.com/1682452/80591955-e42c9380-8a1e-11ea-8488-b082e899a80f.png">

I don't get the notification from step 10: "You have successfully completed...". Will continue investigation.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				);
+				return $message;
+			case 'finish':
+				$message  = __( 'You have successfully completed the notification wizard, good job!', 'wordpress-seo' );
```
Typo: notification wizard => configuration wizard

afercia commented on integration-tests/notifiers/test-class-configuration-notifier.php:
```diff
@@ -0,0 +1,145 @@
+	public function test_notify_when_completed_wizard() {
+		WPSEO_Options::set( 'show_onboarding_notice', false );
+		$notifier = new WPSEO_Configuration_Notifier();
+		$this->assertEquals( 'You have successfully completed the notification wizard, good job!<br/>If you want to double-check your Yoast SEO settings, or change something, you can always <a href="' . get_site_url() . '/wp-admin/admin.php?page=wpseo_configurator">reopen the configuration wizard</a>.', $notifier->notify() );
```
Typo: notification wizard => configuration wizard

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+
+		switch ( $state ) {
+			case 'start':
+				$message  = __( 'Get started quickly with the configuration wizard!', 'wordpress-seo' );
```
This should be escaped.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				$message  = __( 'Get started quickly with the configuration wizard!', 'wordpress-seo' );
+				$message .= '<br/>';
+				$message .= sprintf(
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
```
indentation

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				$message .= '<br/>';
+				$message .= sprintf(
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
+					__( 'We have detected that you have not started this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
```
This should be escaped.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				);
+				return $message;
+			case 'continue':
+				$message  = __( 'The configuration wizard helps you to easily configure your site to have the optimal SEO settings.', 'wordpress-seo' );
```
This should be escaped.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				$message  = __( 'The configuration wizard helps you to easily configure your site to have the optimal SEO settings.', 'wordpress-seo' );
+				$message .= '<br/>';
+				$message .= sprintf(
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
```
indentation

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				$message .= '<br/>';
+				$message .= sprintf(
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
+					__( 'We have detected that you have not finished this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
```
This should be escaped.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
+					__( 'We have detected that you have not started this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
+					'Yoast SEO',
+					'<a href="' . admin_url( '?page=' . self::PAGE_IDENTIFIER ) . '">',
```
URLs should be escaped by using `esc_url( admin_url( ... ) )`

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				/* translators: %1$s resolves to Yoast SEO, %2$s resolves to the starting tag of the link to the wizard, %3$s resolves to the closing link tag */
+					__( 'We have detected that you have not finished this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
+					'Yoast SEO',
+					'<a href="' . admin_url( '?page=' . self::PAGE_IDENTIFIER ) . '">',
```
URLs should be escaped by using `esc_url( admin_url( ... ) )`

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+				/* translators: %1$s expands to Yoast SEO, %2$s is a link start tag to the Onboarding Wizard, %3$s is the link closing tag. */
+					esc_html__( 'If you want to double-check your %1$s settings, or change something, you can always %2$sreopen the configuration wizard%3$s.', 'wordpress-seo' ),
+					'Yoast SEO',
+					'<a href="' . esc_url( admin_url( 'admin.php?page=' . WPSEO_Configuration_Page::PAGE_IDENTIFIER ) ) . '">',
```
URLs should be escaped by using `esc_url( admin_url( ... ) )`

afercia commented on inc/options/class-wpseo-option-wpseo.php:
```diff
@@ -55,6 +55,7 @@ class WPSEO_Option_Wpseo extends WPSEO_Option {
 			],
 			'access_tokens' => [],
 		],
+		'started_configuration_wizard'    => false,
```
Please evaluate with product whether this option should be added to `WPSEO_Tracking_Settings_Data`

afercia replied:
I guess this option should also be validated in some way, see `validate_option()` within this same class.

JesserH replied:
Spoke with @thijsdevalk and this should indeed be moved to `WPSEO_Tracking_Settings_Data`

afercia replied:
I don't think it should be _moved_ 🙂 See for example similar options e.g. `enable_text_link_counter`
The default value is set in this class. Then, it's _added_ to the `WPSEO_Tracking_Settings_Data` class, meaning it's an option to collect for tracking. Not sure it should go in the `anonymous_settings` group (I don't think so) or in the `include_list` group.

afercia commented on admin/config-ui/class-configuration-page.php:
```diff
@@ -214,16 +215,8 @@ private function remove_notification() {
-
+		$note = new Wizard_Notification();
+		$message = $note->get_notification_message( 'continue' );
 		$notification = new Yoast_Notification(
```
Although it appears we don't have a PHPCS rule to enforce it, I think we have some coding style best practice to align this kind of things for better readability.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+ */
+
+/**
+ * Class for handling the different types of wizard notifications
```
Missing period at the end.

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+	 * Gets the notification string.
+	 *
+	 * @param string $state Describes the state of the message that should be returned.
+	 * @return string
```
Missing description

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+		if ( ! $this->show_notification() ) {
+			return $this->re_run_notification();
+		}
+		if (	WPSEO_Options::get( 'started_configuration_wizard' ) ) {
```
typo: tab instead of space

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+	/**
+	 * Returns the notification to re-run the config wizard.
+	 *
+	 * @return Yoast_Notification
```
Actually, it's returning the message string, but...

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+
+		$notification_center = Yoast_Notification_Center::get();
+		$notification_center->add_notification( $notification );
+		return $message;
```
I don't fully understand why it's returning something. Isn't the notification added anyways?

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+	/**
+	 * Returns the notification to start the config wizard for the first time.
+	 *
+	 * @return Yoast_Notification
```
Same as above

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+
+		$notification_center = Yoast_Notification_Center::get();
+		$notification_center->add_notification( $notification );
+		return $message;
```
Same as above

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+	/**
+	 * Returns the notification to continue the config wizard.
+	 *
+	 * @return Yoast_Notification
```
Same as above

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,174 @@
+
+		$notification_center = Yoast_Notification_Center::get();
+		$notification_center->add_notification( $notification );
+		return $message;
```
Same as above

afercia commented on integration-tests/notifiers/test-class-configuration-notifier.php:
```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Unit Test Class.
+ */
```
I'd add a `* @group notifiers`

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,74 @@
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Admin
+ * @since   14.1
```
I think we decided to use `@since` notations only for things like API methods, not in other cases.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 * @return void
+	 */
+	public function listen() {
+		if ( ! $this->show_notification() || ! $this->dismissal_is_triggered() ) {
```
Seems to me `dismiss_get_started` is not used anywhere now? It was used as a URL parameter for the button to _dismiss_ the notification. The notification is not dismissible now.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	/**
+	 * Listens to an argument in the request URL. When triggered just sets the notification to dismissed.
```
This description isn't entirely correct. See also comment below.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 * @return bool True when action has been triggered.
+	 */
+	protected function dismissal_is_triggered() {
+		return filter_input( INPUT_GET, 'dismiss_get_started' ) === '1';
```
It was used as a URL parameter for the button to _dismiss_ the notification. The notification is not dismissible now.

afercia commented:
Question for the Product team:

Should the "success" notification be dismissible? Right now users can only hide it. [On the related issue](https://github.com/Yoast/wordpress-seo/issues/14800#issuecomment-616464749), I don't see an answer to this point:

> Is it correct that only the last notification (3) should be dismissible (besides also being hideable), or should the others also be dismissible? Might also be good to take into account that no other dismissible notifications exist (after wpseo-post-type-archive is removed #14496).

/Cc @JessieHenkes

JesserH commented:
@JessieHenkes Right now the priority of all the three notification is set to `'priority'     => 0.8,`, what is the wanted priority?

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+			return $this->continue_notification();
+		}
+
+		return $this->first_time_notification();
```
If we change the related methods to _not_ return the message string, then we should also remove these 3 returns and just call the methods. The description should be updated as well.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 *
+	 * @var string
+	 */
+	const META_VALUE = 'yes';
```
Depending on feedback on whether the notification should be dismissible, this might be not needed.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 *
+	 * @var string
+	 */
+	const META_NAME = 'wpseo-dismiss-configuration-notice';
```
Depending on feedback on whether the notification should be dismissible, this might be not needed.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+			return;
+		}
+
+		$this->set_dismissed();
```
Depending on feedback on whether the notification should be dismissible, this might be not needed.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 */
+	protected function is_dismissed() {
+		return get_user_meta( get_current_user_id(), self::META_NAME, true ) === self::META_VALUE;
+	}
```
Depending on feedback on whether the notification should be dismissible, this method might be not needed.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 */
+	protected function set_dismissed() {
+		update_user_meta( get_current_user_id(), self::META_NAME, self::META_VALUE );
+	}
```
Depending on feedback on whether the notification should be dismissible, this method might be not needed.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,175 @@
+	 */
+	protected function show_notification() {
+		return $this->show_notification && ! $this->is_dismissed();
+	}
```
Depending on feedback on whether the notification should be dismissible,`&& ! $this->is_dismissed()` might be not needed.

afercia commented on admin/views/tabs/dashboard/dashboard.php:
```diff
@@ -14,8 +14,11 @@
 $notifications_data = Yoast_Notifications::get_template_variables();
 
+$notifier = new WPSEO_Configuration_Notifier();
+$notifier->listen();
```
Could the `listen();` method be called in the class construct?

JesserH replied:
Spoke with @moorscode and this should not be done since constructors are bad practice for side-effects

moorscode replied:
Or phrased differently, it's bad practice to have side-effects in constructors :)
It makes writing tests a lot harder, and also obscures functionality.

afercia commented:
CR & Acceptance 🚧 

Note: some of the things to address depend on feedback from Product team.

JesserH commented:
Spoke with @JessieHenkes about the 2 questions above:

**What should the priority of the three notifications?**
1. First time wizard notice - Highest priority  (Hideable)
2. Continue wizard notice - Highest priority  (Hideable)
3. Finished wizard notice - Low priority         (Hideable)

**Should the finished wizard notice be dismissable?**
No, the notice should not be dismissable since it can be hard to find the configuration wizard without this notice if users want to redo the wizard later on.

JesserH commented:
**IMPORTANT NOTICE
There is no set date when the admin-redesign branch is merged. Therefore, the version for this upgrade routine in `class-upgrade.php` is set to 14.4 because it will probably be merged around that time, however this may change in the future.**

JesserH commented:
After this commit is merged into trunk, [86](https://github.com/Yoast/yoast-test-helper/pull/86) is up for merging as well

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,124 @@
+		$message      = $note->get_notification_message( 'start' );
+		$notification = new Yoast_Notification(
+			$message
+			,
```
this comma shouldn't be in a new line

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	/**
+	 * Adds the re-run notification to the notification centre
```
missing period at the end

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	/**
+	 * Adds the first-time notification to the notification centre
```
missing period at the end

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	/**
+	 * Adds the continue notification to the notification centre
```
missing period at the end

afercia commented on admin/views/tabs/dashboard/dashboard.php:
```diff
@@ -14,8 +14,11 @@
+
 $wpseo_contributors_phrase = sprintf(
-/* translators: %1$s expands to Yoast SEO */
+	/* translators: %1$s expands to Yoast SEO */
```
missng period at the end of the translator comment

afercia commented on inc/class-upgrade.php:
```diff
@@ -53,6 +53,7 @@ public function __construct() {
 			'12.4-RC0'  => 'upgrade_124',
 			'12.8-RC0'  => 'upgrade_128',
 			'13.2-RC0'  => 'upgrade_132',
+			'14.4-RC0'   => 'upgrade_144',
```
alignment

afercia commented on inc/class-upgrade.php:
```diff
@@ -899,4 +900,14 @@ private function migrate_woocommerce_archive_setting_to_shop_page() {
+	private function upgrade_144() {
+		global $wpdb;
+		$wpdb->query( "DELETE FROM $wpdb->usermeta WHERE meta_key = 'wpseo-dismiss-configuration-notice'" );
+	}
```
I think in this file all the `upgrade_*` methods should be grouped together and utility methods should stay at the end of the file.

afercia commented on admin/notifiers/class-configuration-notifier.php:
```diff
@@ -0,0 +1,120 @@
+			$message,
+			[
+				'type'         => Yoast_Notification::WARNING,
+				'id'           => 'wpseo-dismiss-onboarding-notice',
```
The value `wpseo-dismiss-onboarding-notice` is confusing to me, as the notification now can't be "dismissed". See also the other 4 occurrences.

afercia commented:
CR 👍 
Acceptance 🚧 

Getting a bit more familiar with how all this thing works and noticed a few things.

Basically, we're making this notification "permanent". The only things that change are the notification message and priority. T0 make the notification change, it's _removed and then re-added_ at any wizard compeltion state change.

1
I'm not sure I understand why in `admin/config-ui/class-configuration-page.php` there are methods to get / add / remove the notification. I think this class should now only change the values of the `show_onboarding_notice` and `started_configuration_wizard` options. The logic to add / remove the notification should be kept in one place, which is `WPSEO_Configuration_Notifier`.

Right now, the notification is added / removed in two different places. This might explain why there's the need to refresh the dashboard page to actually see the notification change (see point 9 of the test instructions). A refresh should not be necessary.

2
I guess the `started_configuration_wizard` option  should have a default value and for consistency it should be set in the same place where the `show_onboarding_notice` default is set, see `WPSEO_Installation->set_first_install_options()`.

I guess `started_configuration_wizard` should be also added to the tracking data.

3
Minor: the switch statement has no `default` case. Not saying it should be necessarily be added but it's something e should be aware of.

JesserH commented:
The above requested changes by @afercia are implemented, also spoke with @moorscode about the switch case and refactored them into smaller if statements with private functions. Also added tests for those new functions

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,103 @@
+		/* translators: 1: Expands to Yoast SEO, 2: Link start tag to the configuration wizard, 3: Link closing tag. */
+			esc_html__( 'We have detected that you have not started this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
+			'Yoast SEO',
+			'<a href="' . esc_url( admin_url( '?page=' . self::PAGE_IDENTIFIER ) ) . '">',
```
Let's always specify the actual page e.g. `admin_url( 'admin.php?page ...`

afercia commented on admin/class-wizard-notification.php:
```diff
@@ -0,0 +1,103 @@
+		/* translators: 1: Expands to Yoast SEO, 2: Link start tag to the configuration wizard, 3: Link closing tag. */
+			esc_html__( 'We have detected that you have not finished this wizard yet, so we recommend you to %2$sstart the configuration wizard to configure %1$s%3$s.', 'wordpress-seo' ),
+			'Yoast SEO',
+			'<a href="' . esc_url( admin_url( '?page=' . self::PAGE_IDENTIFIER ) ) . '">',
```
Pre-existing issue: Let's always specify the actual page e.g. `admin_url( 'admin.php?page ...`

afercia commented:
CR & acceptance 👍

afercia commented:
👍
